<img align="right" style="padding-left:10px; height: 60%; width: 20%" src="figures/whr.png" >

# World Happiness Report

This case study is based on the 7th World Happiness Report. The first was released in April 2012 in support of a UN High level meeting on “Wellbeing and Happiness: Defining a New Economic Paradigm”. 

That 2012 report presented the available global data on national happiness and reviewed related evidence from the emerging science of happiness, showing that the quality of people’s lives can be coherently, reliably, and validly assessed by a variety of subjective well-being measures, collectively referred to then and in subsequent reports as “happiness.” 

This year’s World Happiness Report focuses on happiness and the community: how happiness has evolved over the past dozen years, with a focus on the technologies, social norms, conflicts and government policies that have driven those changes.

I have downloaded the data from [Chapter 2: Online Data](https://s3.amazonaws.com/happiness-report/2019/Chapter2OnlineData.xls) and filtered out data prior to 2018. The result is available in CSV format, as the next cell shows. _Data Prep Notes:_ The Happiness Score column is from Figure 2.6 in the downloaded report; the other data columns are from Table 2.1 in the same report. If a country wasn't in either list, it wasn't included in the CSV file.

We were first introduced to this dataset in `03-04-world-happiness` and `03-05-world-happiness.ipynb`. We return to the dataset for a deeper analysis!

In [1]:
import pandas as pd
import numpy as np
data1 = pd.read_csv('happiness-report.csv')
data1

Country  Year  HappinessScore  LifeLadder    LogGDP  SocialSupport  \
0    Afghanistan  2018           3.203    2.694303  7.494588       0.507516   
1        Albania  2018           4.719    5.004403  9.412399       0.683592   
2        Algeria  2018           5.211    5.043086  9.557952       0.798651   
3      Argentina  2018           6.086    5.792797  9.809972       0.899912   
4        Armenia  2018           4.559    5.062449  9.119424       0.814449   
..           ...   ...             ...         ...       ...            ...   
131    Venezuela  2018           4.707    5.005663  9.270281       0.886882   
132      Vietnam  2018           5.175    5.295547  8.783416       0.831945   
133        Yemen  2018           3.380    3.057514       NaN       0.789422   
134       Zambia  2018           4.107    4.041488  8.223958       0.717720   
135     Zimbabwe  2018           3.663    3.616480  7.553395       0.775388   

     HealthyLifeExpectancyAtBirth  FreedomToMakeLifeChoices  Generosity  \
0                       52.599998                  0.373536   -0.084888   
1                       68.699997                  0.824212    0.005385   
2                       65.900002                  0.583381   -0.172413   
3                       68.800003                  0.845895   -0.206937   
4                       66.900002                  0.807644   -0.149109   
..                            ...                       ...         ...   
131                     66.500000                  0.610855   -0.176156   
132                     67.900002                  0.909260   -0.039124   
133                     56.700001                  0.552726         NaN   
134                     55.299999                  0.790626    0.036644   
135                     55.599998                  0.762675   -0.038384   

     PerceptionsOfCorruption  PositiveAffect  NegativeAffect  \
0                   0.927606        0.424125        0.404904   
1                   0.899129        0.713300        0.318997   
2                   0.758704        0.591043        0.292946   
3                   0.855255        0.820310        0.320502   
4                   0.676826        0.581488        0.454840   
..                       ...             ...             ...   
131                 0.827560        0.759221        0.373658   
132                 0.808423        0.692222        0.191061   
133                 0.792587        0.461114        0.314870   
134                 0.810731        0.702698        0.350963   
135                 0.844209        0.710119        0.211726   

     ConfidenceInNationalGovernment  
0                          0.364666  
1                          0.435338  
2                               NaN  
3                          0.261352  
4                          0.670828  
..                              ...  
131                        0.260700  
132                             NaN  
133                        0.308151  
134                        0.606715  
135                        0.550508  

[136 rows x 13 columns]

## Some observations:

* **Quality** Some data points just show `NaN` (not a number where there should be one). This is typically done using `dropna()`. 

* **Normalization:** The columns have different ranges: some values are between 0. and 1., Others have different ranges. For example Generosity is between -0.33 and 0.49, and so on. Un-normalized ranges can cause the different features to be under- or over-valued during analysis. The data needs to be preprocessed so as to be uniform. 

* **Pre-normalization:** `LogGDP` is the logarithm of the GDP per capita. Here is a case where the data has gone through Feature Scaling _prior to publication!_ Taking a log of numbers whose range spans multiple orders of magnitude is a common technique for compressing the range. However, it still doesn't span the range [0. 1.] and a bit more Feature Scaling will be required.

### 1. Quality

First, let's try `dropna()`. Out of 136 rows, _we are left with 113_. 

How important is it for data quality to be _absolutely_ sacrosanct in our analysis? To answer this question, we wish to compare `data1` with `data1.dropna()`. Would we lose any important data by doing so? Create a DataFrame `data_na` that shows what we would lose by using `dropna()`.

### Quality Evaluation

Many of the countries in the "to be dropped" list are important in a geopolitical sense. For an analysis of a United Nations dataset called **World Happiness Report**, there are three options:

1. Hold the line on data quality and publish with just 113 countries in the final report,
1. Fill the missing column values with the average value for that column, or
1. (Partial.) Drop the rows with `NaN` for "objective" features such as `LogGDP` and fill the unavailable features with the average values of those features.

### 1a. Choosing the Quality Option

Which of the above quality options would you choose and why? 

_This is a judgment question. No answer is "wrong" or "preferred." How well you make your argument is what's important!_

---

---

---

### 1b. Filled Data

Irrespective of your answer to Q 1a, management has decided to pursue option 3. 

Prepare a DataFrame which has only non-NaN LogGDP rows and the remaining feature values filled with average values of those features. Hint: check out the `np.nanmean()` function.

Finally, reindex the cleaned up DataFrame and rename it `data1_rdy`.

### 2. Normalizing the data

The process of making all columns uniform in scale is referred to as **Feature Scaling**. Many data analysis libraries require it (see [_The Importance of Feature Scaling_](https://scikit-learn.org/stable/auto_examples/preprocessing/plot_scaling_importance.html)). Scikit-Learn offers an extensive library for [preprocessing data](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing). Pay particular attention to `*Scaler()` functions. Also Discretizers such as Binarizer, KBinsDiscretizer and QuantileTransformer.

Use `StandardScaler` to scale all features except `Country`, `Year` and `LogGDP`.

In [ ]:
from sklearn.preprocessing import StandardScaler, RobustScaler
scaler = StandardScaler()
# Fill in...

### 3. Examine scaled features

Examine min, mean and max values of each of the scaled features. Why are the min and max values not 0 and 1 respectively?

**Your Answer**

---

---

---

### 4. Cluster the data into 2 clusters.

The clustering will return a new column with values 0 or 1. We won't yet know what 0 and 1 stand for. Show `data_scaled` along with the returned class values. `kmeans.labels_` gives the class values.

In [ ]:
from sklearn.cluster import KMeans
kmeans = # fill in

### Assigning Cluster Labels

We need to assign labels for values of the clustering cells. The Scikit Learn class LabelEncoder is designed for this purpose. Its usage is shown in the next cell.

In [ ]:
from sklearn.preprocessing import LabelEncoder
from sklearn.utils import column_or_1d

le = LabelEncoder()
le.fit(['unhappy', 'happy'])

def clustered_pd(np_data, np_rows, np_cols, labels, ole):
    return pd.concat([
        pd.DataFrame(np_rows, columns=['Country']), 
        pd.DataFrame(ole.inverse_transform(labels), columns=['Cluster']), 
        pd.DataFrame(np_data, dtype='float32', columns=np_cols)], axis=1)

print (clustered_pd(np_norm, np_rows, np_cols, kmeans.labels_, le).loc[:,['Country', 'Cluster']])

### 5. Assigning Cluster Labels

The labels produced above are **wrong**. Happiness values of 0 are being translated as 'happy'. Fix the code in the above cell to correct this error.

### 6. Make 5 clusters

Based on the above exercise, cluster `data1_scaled` into 5 classes, giving values 1 &hellip; 5 to the appropriate rows.

In [ ]:
from sklearn.cluster import KMeans
n_clusters = 5
kmeans = KMeans(n_clusters=n_clusters, random_state=0).fit(np_norm)
kmeans.labels_

## Automatically assigning labels

Replace arbitrary numeric labels with descriptive numerical labels for categories. Sometimes it is possible to use some column values to give us a hint about cluster identity. For example, in the data above, Happiness Score could be used to label the rows.

Change the happiness values to be 'sad', 'down', 'ok', 'up', 'happy'.

# When you're done, submit the notebook

1. **Run all the cells in order.**

2. Submit the notebook by saving it as PDF. 
    * In the cluster environment, it's File | Print (Save as PDF) and submit to [Gradescope](https://www.gradescope.com/courses/182658)<sup>&dagger;</sup>, 
    * On other versions, it may be File | Download As (PDF) and then submit to [Gradescope](https://www.gradescope.com/courses/182658)<sup>&dagger;</sup>.

<sup>&dagger;</sup>To submit to Gradescope, log into the website, add course 9W7PW3 (if not already added) and submit. The assignment name should match the name of this notebook.

![The end](https://live.staticflickr.com/32/89187454_3ae6aded89_b.jpg)